### ISSUE TITLE
Supernets with varying block sizes

### ISSUE THREAD

**Issue by gabikadlecova** [whittle-org/whittle#137]
tl;dr support supernets with varying block sizes (`intermediate_size` and `n_heads` can be lists)

**Is your feature request related to a problem? Please describe.**
When converting HW-GPT-Bench subnets to whittle, you need to first initialize the supernet, and then call `set_sub_network` to set the correct intermediate sizes and n_heads.

This means that if we use only the subnetwork, we get a lot of unused weights. Ideally, the subnet should be initialized "as a supernet", i.e. setting `n_embd` to the smaller embedding size.

While we can set the embedding size, we currently cannot work with `intermediate_size` as a list due to how `config.intermediate_size` is used inside the MLPs:
https://github.com/whittle-org/whittle/blob/3b18ba58a60ed0266438b67b0cfc272a291c7cb9/whittle/models/gpt/blocks/mlp.py#L14

The same applies for n_head (currently, it is the same for all blocks).

**Describe the solution you'd like**
Enable lists of ints for `config.intermediate_size` and `config.n_head`

**Comment by gabikadlecova** [comment]
I have a fix ready, I just want to write a test validating that the supernet produces the same output as a subnet created via `set_sub_network`.

**Comment by aaronkl** [comment]
Hey Gabi, thanks a lot for your interest. Just to make sure I understand you correctly, you would like to specify a different number of heads / intermediate size per layer? The `set_sub_network` function actually expects a list of integers for each layer (see [here](https://github.com/whittle-org/whittle/blob/3b18ba58a60ed0266438b67b0cfc272a291c7cb9/whittle/models/gpt/model.py#L175)).

**Comment by gabikadlecova** [comment]
Hi @aaronkl , yes, the `set_sub_network` enables that. But initializing a supernet with uneven block sizes does not seem to be possible

I implemented a fix in the linked branch:
https://github.com/whittle-org/whittle/compare/main...flexible-intermediate-sizes

**Comment by gabikadlecova** [comment]
I added tests in `tests/test_extract.py` to verify the initialized supernet behaves the same as a subnet set via `set_sub_networks`.

https://github.com/whittle-org/whittle/commit/71cbb9b520e464b501bd38af5d7869fba61d59a4

However, there are two problems

a) `litgpt` does not support lists for `n_head` or `intermediate_size` in Config. Therefore, a call to `super().__init__(config, ...)` will lead to a crash.
I solved it in a hacky way - I set both fields to a nonsensical value before passing them to super init, since we reimplement all submodules anyway. Another solution would be to introduce a field with a different name. What do you think would be the best way how to solve it?

b) Some field in `CausalSelfAttention` are computed dynamically in some cases:
- `sub_network_query_groups` if the `n_head` is not divisible by `n_query_groups`
- `attention_scores_scalar` if nonzero
- head_size if `fix_head_size == False`
This means that they should also be represented as a list in the config. These are the cases that I traced in the code, is there anything else?

**Comment by rheasukthanker** [comment]
Thanks for raising this point @gabikadlecova. While I agree that in principle, varying heads and intermediate sizes per layer are nice to have, I think we used fixed dimension as litgpt config sets intermediate dim and heads as a fixed integer. Also in general a search space with fixed embed dim and heads per layer is more aligned with handcrafted transformer architectures, and varying dims per layer are a bit unnatural. I think we should discuss this issue in the weekly library synch meeting @aaronkl and @gabikadlecova

**Comment by gabikadlecova** [comment]
@rheasukthanker makes sense, sure!

While looking at the code, I noticed that `extract_sub_network` works just for `GptNeoxMLP` and not for `LLaMAMLP` and `GemmaMLP`. Should I create a separate issue, or is it not a priority right now?

**Comment by rheasukthanker** [comment]
@gabikadlecova , thanks for raising that point, I somehow never noticed that. Yes, please create a separate issue and PR :)

**Comment by aaronkl** [comment]
@gabikadlecova Could you tell more about your use-case? To add to Rhea's point, we ran some preliminary experiments some time ago, and found that varying the number of heads / intermediate size per layer didn't yield much benefit.
